### PR TITLE
Containerize Gantry

### PR DESCRIPTION
**Pull request by cmelone**
Added a Dockerfile and some docs about how to run the app. Also moved away from using Spack to handle dependencies as we can use a lightweight/distroless container that allows for a small image size when installing via pip.

### CODE REVIEW

**Comment by alecbcs** [comment]
If possible it'd be nice to keep the Spack environment in the repo for development purposes.

**Comment by cmelone** [comment]
Happy to do that, I'll put it back in

**Comment by cmelone** [comment]
I put the environment file back in, but I prefer leaving the auto-activation out of `.envrc` for now as I'm not using Spack in my own environment atm

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /venv/bin/pip install --upgrade pip setuptools wheel
+
+# Build the virtualenv as a separate step: Only re-execute this step when requirements.txt changes
+FROM build AS build-venv
+COPY requirements.txt /requirements.txt
+RUN /venv/bin/pip install --disable-pip-version-check -r /requirements.txt
+
+# Copy the virtualenv into a distroless image
+FROM gcr.io/distroless/python3-debian12:nonroot
+COPY --from=build-venv /venv /venv
```
Comment (alecbcs) [inline Dockerfile]:
```suggestion
COPY requirements.txt /requirements.txt
RUN /venv/bin/pip install --disable-pip-version-check -r /requirements.txt

# Copy the virtualenv into a distroless image
FROM gcr.io/distroless/python3-debian12:nonroot
COPY --from=build /venv /venv
```

As long as the pip install and the apt-get are in separate `RUN` directives they'll be applied in different layers and should cache just about as well without the intermediate container.

File: .envrc [hunk 2]
```diff
@@ -1,14 +1,3 @@
...
-# Load Development Spack Environment (If Spack is installed.)
-#
-# Run 'direnv allow' from within the cloned repository to automatically
-# load the spack environment when you enter the directory.
-#------------------------------------------------------------------------
-if type spack &>/dev/null; then
-    . $SPACK_ROOT/share/spack/setup-env.sh
-    spack env activate -d .
-fi
-
```
Comment (alecbcs) [inline .envrc]: If you remove/don't have a spack.lock file does activating the environment actually mess with your install through pip? Or is this purely to speed up the direnv execution?

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,3 @@
 __pycache__
 .env
-spack.lock
-.spack-env
```
Comment (alecbcs) [inline .gitignore]: It'd be nice to keep these in the `.gitignore` for other folks who might be using Spack.

**Review by alecbcs — CHANGES_REQUESTED** [review]
Looks great! Just a few quick suggestions.

File: Dockerfile (hunk 1 above)
Comment (cmelone) [reply Dockerfile]: good catch, addressed via https://github.com/spack/spack-gantry/pull/10/commits/4ee0db44526f1bc95b03a2e6c2b7cc1f9d340332

File: .envrc (hunk 2 above)
Comment (cmelone) [reply .envrc]: addressed via https://github.com/spack/spack-gantry/pull/10/commits/f54656223688f4be216d69fea5493150f810c77b

**Review by alecbcs — APPROVED** [review]
Looks good to me!

**Pull request merged**